middleware/step_build.go

```diff
@@ -0,0 +1,182 @@
+package middleware
+
+import "context"
+
+// BuildInput provides the input parameters for the BuildMiddleware to consume.
+// BuildMiddleware may modify the Request value before forwarding the input
+// along to the next BuildHandler.
+type BuildInput struct {
+	Request interface{}
+}
+
+// BuildOutput provides the result returned by the next BuildHandler.
+type BuildOutput struct {
+	Result interface{}
+}
+
+// BuildHandler provides the interface for the next handler the
+// BuildMiddleware will call in the middleware chain.
+type BuildHandler interface {
+	HandleBuild(ctx context.Context, in BuildInput) (
+		out BuildOutput, err error,
+	)
+}
+
+// BuildMiddleware provides the interface for middleware specific to the
+// serialize step. Delegates to the next BuildHandler for further
+// processing.
+type BuildMiddleware interface {
+	// Unique ID for the middleware in theBuildStep. The step does not allow
+	// duplicate IDs.
+	ID() string
+
+	// Invokes the middleware behavior which must delegate to the next handler
+	// for the middleware chain to continue. The method must return a result or
+	// error to its caller.
+	HandleBuild(ctx context.Context, in BuildInput, next BuildHandler) (
+		out BuildOutput, err error,
+	)
+}
+
+// BuildMiddlewareFunc returns a BuildMiddleware with the unique ID provided,
+// and the func to be invoked.
+func BuildMiddlewareFunc(id string, fn func(context.Context, BuildInput, BuildHandler) (BuildOutput, error)) BuildMiddleware {
```

Comment:
Curious on your thoughts for not exposing the `buildMiddlewareFunc` and it's members publically? I know it's kind of hit or miss around the community on whether functions/methods in libraries should return concrete vs interface types.

Reply:
Main reason for not exposing the struct was to force usage of the function, to prevent any accidental miss reading of the struct's requirements and incorrect thinking ID, or Fn may be optional.